### ISSUE TITLE
system.clients is missing: hostname, ssl_cipher_suite, ssl_enabled, ssl_protocol

### ISSUE THREAD

**Issue by jul-stas** [scylladb/scylladb#9216]
This is nothing new; i.e. these columns were intentionally left to be implemented later.

**Comment by mykaul** [comment]
I think it was fixed by https://github.com/scylladb/scylladb/issues/6946

**Comment by mykaul** [comment]
> I think it was fixed by #6946

It actually wasn't - we have tests that currently are skipped as there are missing fields yet (see the relevant PR for details), so need to keep this open.

**Comment by DanielHe4rt** [comment]
Any news about this issue? I stepped into TLS/SSL recently and saw that we still don't have this implemented. It would be really good to demonstrate the feature with some type of logging when retrieving by `select * from system.clients;` query.

I found this merge https://github.com/scylladb/scylladb/commit/fb24bccf046d319e6f27ed6a6dc0f9079d19d741 but nothing more related to it...

**Comment by mykaul** [comment]
> Any news about this issue? I stepped into TLS/SSL recently and saw that we still don't have this implemented. It would be really good to demonstrate the feature with some type of logging when retrieving by `select * from system.clients;` query.

Contributions are welcome. This was not implemented yet.

> 
> I found this merge [fb24bcc](https://github.com/scylladb/scylladb/commit/fb24bccf046d319e6f27ed6a6dc0f9079d19d741) but nothing more related to it...

which is very clear:
```
What is still missing:
* SSL params (I'll happily get some advice here)
```

**Comment by mykaul** [comment]
Relevant PR that *did not* implement this - https://github.com/scylladb/scylladb/pull/7349

**Comment by mykaul** [comment]
CC @ptrsmrn - I've mentioned this item in the call today.

**Comment by ptrsmrn** [comment]
Missing columns are present in `Scylla version 2025.2.0~dev-0.20250218.aa8c27b87200`:
```
cqlsh> DESC system.clients ;

CREATE TABLE system.clients (
...
    hostname text,
...
    ssl_cipher_suite text,
    ssl_enabled boolean,
    ssl_protocol text,
...
```
But quick check shows they're set to NULLs:
```
cqlsh> select * from system.clients ;

@ Row 1
------------------+------------------------
 address          | 127.0.0.1
 port             | 59348
 client_type      | cql
 connection_stage | READY
 driver_name      | DataStax Python Driver
 driver_version   | 3.26.0
 hostname         | null
 protocol_version | 4
 scheduling_group | main
 shard_id         | 0
 ssl_cipher_suite | null
 ssl_enabled      | null
 ssl_protocol     | null
 username         | anonymous
```
So I think the issue is not that the columns are not there, but rather that they are not set - I don't see any place in the source code that sets them.

**Comment by nyh** [comment]
@ptrsmrn I have two questions:
1. One of the comments above suggests we already have tests for these fields but they are skipped. Do you know what those are (in dtest? or what?). Did you test that your fix makes them pass? Do you plan to send a patch to un-skip them?
2. This issue and the PR do not even mention the word "Cassandra"... Are these fields provided by Cassandra? Do we need to implement them for Cassandra compatibility? Please note that test/cqlpy/run-cassandra also runs Cassandra by default without SSL and when the "--ssl" option is passed, runs Cassandra **with** SSL, which makes it very easy to test (in manual runs, not CI) both the no-ssl and with-ssl scenarios.

**Comment by ptrsmrn** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/124208650?s=64&amp;v=4">@ptrsmrn](https://github.com/ptrsmrn) I have two questions:
> 
> 1. One of the comments above suggests we already have tests for these fields but they are skipped. Do you know what those are (in dtest? or what?). Did you test that your fix makes them pass? Do you plan to send a patch to un-skip them?

I didn't notice, and thanks for pointing that out, but there's a bunch of dtests blocked on this issue - https://github.com/scylladb/scylla-dtest/blob/next/cql_system_clients_test.py#L449, which makes me think I could drop my test.py testcases I'm trying to introduce in https://github.com/scylladb/scylladb/pull/22961? But maybe better to leave the ones in test.py and remove dtests, because: (1) we're trying to migrate dtests to test.py, (2) dtests are harder to maintain.

> 2. This issue and the PR do not even mention the word "Cassandra"... Are these fields provided by Cassandra? Do we need to implement them for Cassandra compatibility? Please note that test/cqlpy/run-cassandra also runs Cassandra by default without SSL and when the "--ssl" option is passed, runs Cassandra **with** SSL, which makes it very easy to test (in manual runs, not CI) both the no-ssl and with-ssl scenarios.

This is interesting, `system.clients` table does not exist in C*. When I run the testcase modified in my PR against C* I get:
```
[piotrsmaron@fedora cqlpy]$ ./run-cassandra -s --ssl test_ssl.py::test_tls_versions 
Cassandra is: /home/piotrsmaron/Downloads/apache-cassandra-5.0-rc2-bin/apache-cassandra-5.0-rc2/bin/cassandra.
...
>                   table_result = session.execute(f"SELECT * FROM system.clients")
...
E   cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="table clients does not exist"
```
So it's either in a different table or I need to run my test only against ScyllaDB.

**Comment by nyh** [comment]
> > 1. One of the comments above suggests we already have tests for these fields but they are skipped. Do you know what those are (in dtest? or what?). Did you test that your fix makes them pass? Do you plan to send a patch to un-skip them?
> 
> I didn't notice, and thanks for pointing that out, but there's a bunch of dtests blocked on this issue - https://github.com/scylladb/scylla-dtest/blob/next/cql_system_clients_test.py#L449, which makes me think I could drop my test.py testcases I'm trying to introduce in [#22961](https://github.com/scylladb/scylladb/pull/22961)?

Please don't. I think the test.py tests are the important ones in the long run, and we should have them and not rely on dtest.

The test.py tests is something a real developer can use without wasting hours of their lives, and importantly, it's something you can also run against Cassandra.

> But maybe better to leave the ones in test.py and remove dtests, because: (1) we're trying to migrate dtests to test.py, (2) dtests are harder to maintain.

Since we have a bazillion tests in dtest, I don't think that removing a couple will make any difference. So feel free to just leave them.

> 
> > 2. This issue and the PR do not even mention the word "Cassandra"... Are these fields provided by Cassandra? Do we need to implement them for Cassandra compatibility? Please note that test/cqlpy/run-cassandra also runs Cassandra by default without SSL and when the "--ssl" option is passed, runs Cassandra **with** SSL, which makes it very easy to test (in manual runs, not CI) both the no-ssl and with-ssl scenarios.
> 
> This is interesting, `system.clients` table does not exist in C*. When I run the testcase modified in my PR against C* I get:
> 
> ```
> [piotrsmaron@fedora cqlpy]$ ./run-cassandra -s --ssl test_ssl.py::test_tls_versions 
> Cassandra is: /home/piotrsmaron/Downloads/apache-cassandra-5.0-rc2-bin/apache-cassandra-5.0-rc2/bin/cassandra.
> ...
> >                   table_result = session.execute(f"SELECT * FROM system.clients")
> ...
> E   cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="table clients does not exist"
> ```
> 
> So it's either in a different table or I need to run my test only against ScyllaDB.

Thanks for looking at this. I believe that Cassandra did have system.clients in the past (?) but now according to 
https://cassandra.apache.org/doc/stable/cassandra/new/virtualtables.html it's called `system_views.clients`.

It's a good question why it is fine that our name is different from Cassandra :-( 

If our schema is the same and just the name of the table is different, and we are fine with this different, you can add a loop that tries both names :-( Or does something like
```
  ks = 'system' if is_scylla(cql) else 'system_views`
```

**Comment by ptrsmrn** [comment]
OK, I'll keep my newly added test.py testcases. We'll see if dtests need to be adjust - they'll start to run once this issue is closed.
I can make the test.py testcases run for both Scylla and Cassandra, but there are slight differences - Cassandra's `system_views.clients` has these values:
1. hostname='localhost' vs mine hostname='127.0.0.1'
2. ssl_cipher_suite='TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256',
3. ssl_enabled=True,
4. ssl_protocol='TLSv1.2' vs mine ssl_protocol='TLS1.2' (the lack of the 'v' character).

So I'll just need to accept one or the other version.

**Comment by mykaul** [comment]
> 4\. ssl_protocol='TLSv1.2' vs mine ssl_protocol='TLS1.2' (the lack of the 'v' character).

I wonder why you are not seeing TLS 1.3 (with or without the v...)

**Comment by nyh** [comment]
https://github.com/scylladb/scylladb/issues/23264 also asks to improve the documentation about system.clients and the meaning of its different columns.

**Comment by mykaul** [comment]
Do we need to improve test_system_clients to accommodate the changes?

**Comment by ptrsmrn** [comment]
> Do we need to improve test_system_clients to accommodate the changes?

@mykaul are you referring to dtests? I wasn't aware there are dtests for system.clients. If they fail, I'll adjust them, but the new functionality is tested in test.py, and I don't think it needs to be tested again in dtests.

**Comment by mykaul** [comment]
> > Do we need to improve test_system_clients to accommodate the changes?
> 
> [@mykaul](https://github.com/mykaul) are you referring to dtests? I wasn't aware there are dtests for system.clients. If they fail, I'll adjust them, but the new functionality is tested in test.py, and I don't think it needs to be tested again in dtests.

Well, if we do a great job in test.py vs. existing dtest (see https://github.com/scylladb/scylla-dtest/blob/61a67fb747d4eb0ac0e1a541a1095cf0f9ebe101/cql_system_clients_test.py#L20 and friends), then we clearly should remove this dtest!

**Comment by ptrsmrn** [comment]
> > > Do we need to improve test_system_clients to accommodate the changes?
> > 
> > 
> > [@mykaul](https://github.com/mykaul) are you referring to dtests? I wasn't aware there are dtests for system.clients. If they fail, I'll adjust them, but the new functionality is tested in test.py, and I don't think it needs to be tested again in dtests.
> 
> Well, if we do a great job in test.py vs. existing dtest (see [scylladb/scylla-dtest@`61a67fb`/cql_system_clients_test.py#L20](https://github.com/scylladb/scylla-dtest/blob/61a67fb747d4eb0ac0e1a541a1095cf0f9ebe101/cql_system_clients_test.py#L20) and friends), then we clearly should remove this dtest!

The linked dtests are way more comprehensive than what we have in test.py at the moment. In test.py we only test some corner cases (and I added 2 tests for another (tls) corner cases). 
So the task to migrate dtests to test.py is something to definitely consider, but as a separate effort.

**Comment by mykaul** [comment]
@ptrsmrn - small ask - can we also add compression used, or is that requires changing the schema and so forth?

**Comment by ptrsmrn** [comment]
> [@ptrsmrn](https://github.com/ptrsmrn) - small ask - can we also add compression used, or is that requires changing the schema and so forth?

@mykaul that probably requires extending seastar, the same way I did it here: https://github.com/scylladb/seastar/pull/2651/files, but overall shouldn't be complicated I guess. 
And yes, the system.clients table doesn't have tls_compression column, so schema alteration is needed.

**Comment by mykaul** [comment]
> > [@ptrsmrn](https://github.com/ptrsmrn) - small ask - can we also add compression used, or is that requires changing the schema and so forth?
> 
> [@mykaul](https://github.com/mykaul) that probably requires extending seastar, the same way I did it here: https://github.com/scylladb/seastar/pull/2651/files, but overall shouldn't be complicated I guess. And yes, the system.clients table doesn't have tls_compression column, so schema alteration is needed.

I doubt TLS is going to be compressed. I was referring to the CQL Snappy/LZ4/None

**Comment by ptrsmrn** [comment]
> > > [@ptrsmrn](https://github.com/ptrsmrn) - small ask - can we also add compression used, or is that requires changing the schema and so forth?
> > 
> > 
> > [@mykaul](https://github.com/mykaul) that probably requires extending seastar, the same way I did it here: [scylladb/seastar#2651 (files)](https://github.com/scylladb/seastar/pull/2651/files), but overall shouldn't be complicated I guess. And yes, the system.clients table doesn't have tls_compression column, so schema alteration is needed.
> 
> I doubt TLS is going to be compressed. I was referring to the CQL Snappy/LZ4/None

@mykaul OK, I can look into that. At least it's going to be easier from the process perspective, if no seastar changes are required.

**Comment by ptrsmrn** [comment]
It's been decided not to populate the `hostname` column. Cassandra does a costly rDNS to have it filled in, and we don't want this performance penalty - if customers want to do it, they can to it themselves. The discussion: https://github.com/scylladb/scylladb/pull/22961#discussion_r2328730099
